### DISCUSSION TITLE
How to give the desired position and velocity together by Setpoint?

### DISCUSSION THREAD

**Discussion by Residualstress** [bitcraze/discussions#1667] [Q&A] [answered]
Hi, can the drone fly to the desired position with a  given velocity?
I know how to use Absolute mode (modeAbs) and Velocity mode (modeVelocity) in the Setpoint to control the drone. But when I used position control, the drone moved too fast. Any hints on it? Thank you.

**Comment by gemenerik** [comment]
I recommend using the high level commander for something like this. `crtpCommanderHighLevelGoTo2` has a duration input which will affect the velocity. If you want a constant velocity, you can use linear mode (recently introduced in https://github.com/bitcraze/crazyflie-firmware/pull/1410). How you do this exactly depends on where you are implementing it (onboard or from the lib for example).

**Reply by Residualstress** [reply]
Thanks for the suggestion. Actually, I used `setHoverSetpoint` to control the drone in App layer. I learned this method from [the push demo](https://github.com/bitcraze/crazyflie-firmware/blob/7c64c98060db7622c77c808d97797abb59ef9e9f/examples/demos/app_push_demo/src/push.c#L47C1-L47C95). Can I solve this problem with using `setHoverSetpoint`?
My idea is that there should be a parameter to set the velocity when we use Absolute mode (modeAbs). So if I can change that value, the velocity could be controled.

**Reply by gemenerik** [reply, marked as answer]
The controllers don’t support controlling both position and velocity at the same time; they can only handle one mode for each dimension. The `setHoverSetpoint` function you’re using creates setpoints fed into the controller. To handle simultaneous velocity and position goals, a higher-level layer like the High Level Commander is needed to calculate and provide setpoints accordingly.
